### PR TITLE
handle race condition: close handled before open

### LINKED ISSUE #266: ConnectStart/ConnectDone() called multiple times, causing panic
Stack trace from @austinlparker:

```
11:19 AM
11:19 AM
goroutine 592 [running]:
11:19 AM
go.opentelemetry.io/plugin/httptrace.(*clientTracer).close(0xc00055ec00, 0x828607, 0xc)
11:19 AM
	/app/pkg/mod/go.opentelemetry.io@v0.0.0-20191023164856-8cb378678945/plugin/httptrace/clienttrace.go:79 +0x21f
11:19 AM
go.opentelemetry.io/plugin/httptrace.(*clientTracer).connectDone(...)
11:19 AM
	/app/pkg/mod/go.opentelemetry.io@v0.0.0-20191023164856-8cb378678945/plugin/httptrace/clienttrace.go:125
11:19 AM
net.(*sysDialer).dialSingle.func1(0xc000493cc0, 0xc0004d0f80, 0xc000512bc0, 0xe, 0xc0001b9e00)
11:19 AM
	/tmp/go/src/net/dial.go:571 +0x6b
11:19 AM
net.(*sysDialer).dialSingle(0xc0004d0f80, 0x8b1900, 0xc000483f00, 0x8ad2e0, 0xc000485e30, 0x8b4140, 0xc00031e008, 0x0, 0x0)
11:19 AM
	/tmp/go/src/net/dial.go:594 +0x2fa
11:19 AM
net.(*sysDialer).dialSerial(0xc0004d0f80, 0x8b1900, 0xc000483f00, 0xc000529770, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0)
11:19 AM
	/tmp/go/src/net/dial.go:546 +0x221
11:19 AM
net.(*sysDialer).dialParallel.func1(0x8b1900, 0xc000483f00, 0xc000529701)
11:19 AM
	/tmp/go/src/net/dial.go:467 +0x95
11:19 AM
created by net.(*sysDialer).dialParallel
11:19 AM
	/tmp/go/src/net/dial.go:482 +0x25a
11:19 AM
exit status 2
```

My analysis:
```
Okay, so this happens in https://github.com/open-telemetry/opentelemetry-go/blob/master/plugin/httptrace/clienttrace.go#L80

this happens when the close() call happens _before_ the open() has happened. because concurrency and goroutines and delayed reporting of HTTP events :/

One solution is to just... not panic and drop the span. I prefer that one. but the other is to actually block and wait for the open span to happen before the close span call is run, if it isn't in the map. *cry*

this used to be worse, it would close the last span reported, rather than closing the correct span. so if run in the wrong order, you'd see open-x, open-y, close-y (meant to close-x), close-??? (uh oh) happening. because of the ordering not being open-close-open-close.

we now deal correctly with open-open-close-open-close-close, but not with close-open.
```

### PR DESCRIPTION
**Pull request by rghetia**
closes #266

### CODE REVIEW

File: plugin/httptrace/clienttrace.go [hunk 1]
```diff
@@ -77,7 +82,8 @@ func (ct *clientTracer) close(name string) {
 		s.End()
 		delete(ct.levels, name)
 	} else {
-		panic(fmt.Sprintf("failed to find span %s in levels.", name))
+		// open is not finished before close is called.
+		ct.levels[name] = s
```
Comment (krnowak) [inline plugin/httptrace/clienttrace.go]: I don't get this assignment, what is `s` here? A `(type: nil, value: nil)` interface? A `(type: trace.Span, value: nil)` interface? Whatever it is, it feels like we are putting a bogus value in the map, and I can't see that we check anywhere if the value we got from the map is valid.

Comment (paivagustavo) [reply plugin/httptrace/clienttrace.go]: Agree with @krnowak, I think we need some test case to validate this.

Comment (rghetia) [reply plugin/httptrace/clienttrace.go]: yes. I am going to close this.

**Comment by rghetia** [comment]
the solution doesn't work.

File: plugin/httptrace/clienttrace.go (hunk 1 above)
Comment (rghetia) [reply plugin/httptrace/clienttrace.go]: I have added test for race condition but race condition is not happening even without the fix.

**Comment by rghetia** [comment]
@krnowak , @paivagustavo , @jmacd PTAL.

File: plugin/httptrace/clienttrace.go (hunk 1 above)
Comment (paivagustavo) [reply plugin/httptrace/clienttrace.go]: > I have added a test for race condition but race condition is not happening even without the fix.

The only way I could repro this issue was by reusing the same http.Request:
```go
req, _ := http.NewRequest("GET", ts.URL, nil)
_, req = httptrace.W3C(ctx, req)
for i := 0; i < 100; i++ {
	res, err := client.Do(req)
	if err != nil {
		return err
	}
	res.Body.Close()
}
```

@lizthegrey are there any chances you're doing this?

Comment (lizthegrey) [reply plugin/httptrace/clienttrace.go]: Now that you mention it, yes, yes my code was doing that.

Comment (lizthegrey) [reply plugin/httptrace/clienttrace.go]: I figured it out. See comments on PR.

**Comment by lizthegrey** [comment]
A more detailed diagnosis: this is _intentional_ behavior by golang to call ConnectDone() more than once.

https://github.com/golang/go/blob/master/src/net/dial.go#L424

`dialParallel` kicks off more than one `startRacer()` forked from the same `ctx`, meaning the `ctx.Value(nettrace.TraceKey{}).(*nettrace.Trace)` is shared. we can then wind up calling:

```
		if trace.ConnectStart != nil {
			trace.ConnectStart(sd.network, raStr)
		}
		if trace.ConnectDone != nil {
			defer func() { trace.ConnectDone(sd.network, raStr, err) }()
		}
```

In particular, `ConnectStart()` and `ConnectDone()` are each run in pairs, but if `ConnectStart(),ConnectStart(),ConnectDone(),ConnectDone()` happens we'll choke. We add to the map the first time, overwrite the entry the second time, and then first Done removes from map, and the second Done panics.

`ConnectStart(),ConnectDone(),ConnectStart(),ConnectDone()` behaves correctly because the map insertion/deletion is paired correctly.

**Comment by paivagustavo** [comment]
Yes @lizthegrey, I have came to the same conclusion.

It looks like from the trace you posted on #266, the problem lies when the `net.Dialer` tries to dial connections in parallel. This happens when a name is mapped to both ipv4 and ipv6  (I needed to add an entry on my `/etc/hosts` with the `::1` ipv6 mapping to localhost to repro this bug).

When this parallel dialer is called, the second `http.connect` is activated if the first one didn't answer in some time (300ms is the default), so we receive to `connectStart` hooks, but with different addresses. 

I suggest us to change the span name of `connectStart` to contain the address, that would result in a span that looks similar to:
image

**Comment by lizthegrey** [comment]
> I suggest us to change the span name of `connectStart` to contain the address, that would result in a span that looks similar to:
> image
It feels like the address connected to should be an attribute of the span rather than in its name per se, but agreed that we can use it in the map key to disambiguate the separate racing connect calls :)

over to #266 perhaps?

**Comment by lizthegrey** [comment]
Superseded by https://github.com/open-telemetry/opentelemetry-go/pull/283

**Comment by rghetia** [comment]
Superseded by #283

**Pull request closed without merging**